armi/reactor/composites.py

```diff
@@ -2300,6 +2321,254 @@ def getBoundingIndices(self):
+
+        componentsWithThisMat = []
+        for c in self.getComponents():
+            if c.getProperties().getName() == materialName:
+                componentsWithThisMat.append(c)
+        return componentsWithThisMat
+
+    def hasComponents(self, typeSpec):
+        """
+        Return true if all of the named components exist on this block.
+
+        Parameters
+        ----------
+        typeSpec : Flags or iterable of Flags
+            Component types to check for. If None, will check for any components
+        """
+        # Wrap the typeSpec in a tuple if we got a scalar
+        try:
+            iterator = iter(typeSpec)
+        except TypeError:
+            typeSpec = (typeSpec,)
+
+        for t in typeSpec:
+            if not self.getComponents(t):
+                return False
+        return True
+
+    def getComponentByName(self, name):
+        """
+        Gets a particular component from this block, based on its name
+
+        Parameters
+        ----------
+        name : str
+            The blueprint name of the component to return
+        """
+        components = [c for c in self.getComponents() if c.name == name]
+        nComp = len(components)
+        if nComp == 0:
+            return None
+        elif nComp > 1:
+            raise ValueError(
+                "More than one component named '{}' in {}".format(self, name)
+            )
+        else:
+            return components[0]
+
+    def getComponent(self, typeSpec, exact=False, returnNull=False, quiet=False):
+        """
+        Gets a particular component from this block.
+
+        Parameters
+        ----------
+        typeSpec : flags.Flags or list of Flags
+            The type specification of the component to return
+
+        exact : boolean, optional
+            Demand that the component flags be exactly equal to the typespec. Default: False
+
+        quiet : boolean, optional
+            Warn if the component is not found. Default: False
+
+        Careful with multiple similar names in one block
+
+        Returns
+        -------
+        Component : The component that matches the critera or None
+
+        """
+        results = self.getComponents(typeSpec, exact=exact)
+        if len(results) == 1:
+            return results[0]
+        elif not results:
+            if not quiet:
+                runLog.warning(
+                    "No component matched {0} in {1}. Returning None".format(
+                        typeSpec, self
+                    ),
+                    single=True,
+                    label="None component returned instead of {0}".format(typeSpec),
+                )
+            return None
+        else:
+            raise ValueError(
+                "Multiple components match in {} match typeSpec {}: {}".format(
+                    self, typeSpec, results
+                )
+            )
+
+    def iterComponents(self, typeSpec: TypeSpec = None, exact=False):
```

Comment:
Surprised that get component is defined, but iter component is not. Might be worthwhile to add a comment describing why.

Reply:
The reason is because the implementation of getComponent is identical whether you're on a Composite or a Leaf whereas iterComponents is different for composite (loop over children) vs. leaf (yield self). I will try to make that more clear in the class docstring but I'm doing that to follow what (I think) the design pattern suggests.

Reply:
Okay cool. I didn't realize that getComponent will actually raise not implemented since getComponents is not implemented. Its just a thin wrapper.

Might be worth considering moving iterComponents up to the other notImplemented methods like getComponents.

Reply:
good idea on the move. Done.